spring-session/src/main/java/org/springframework/session/data/gemfire/GemFireOperationsSessionRepository.java

```diff
@@ -0,0 +1,352 @@
+			GemFireSession session = new GemFireSession();
+			session.setMaxInactiveIntervalInSeconds(maxInactiveIntervalInSeconds);
+			return session;
+		}
+
+		public static GemFireSession from(ExpiringSession expiringSession) {
+			GemFireSession session = new GemFireSession(expiringSession);
+			session.setLastAccessedTime(System.currentTimeMillis());
+			return session;
+		}
+
+		private String validateId(String id) {
+			Assert.hasText(id, "The Session ID must be specified");
+			return id;
+		}
+
+		@Override
+		public String getId() {
+			return id;
+		}
+
+		@Override
+		public void setAttribute(String attributeName, Object attributeValue) {
+			if (attributeValue != null) {
+				sessionAttributes.put(attributeName, attributeValue);
+			}
+			else {
+				removeAttribute(attributeName);
+			}
+		}
+
+		@Override
+		public void removeAttribute(String attributeName) {
+			sessionAttributes.remove(attributeName);
+		}
+
+		@Override
+		public <T> T getAttribute(String attributeName) {
+			return (T) sessionAttributes.get(attributeName);
+		}
+
+		@Override
+		public Set<String> getAttributeNames() {
+			return Collections.unmodifiableSet(sessionAttributes.keySet());
+		}
+
+		@Override
+		public long getCreationTime() {
+			return creationTime;
+		}
+
+		@Override
+		public boolean isExpired() {
+			long lastAccessedTime = getLastAccessedTime();
+			long maxInactiveIntervalInSeconds = getMaxInactiveIntervalInSeconds();
+
+			return (maxInactiveIntervalInSeconds >= 0
+				&& (System.currentTimeMillis() - TimeUnit.SECONDS.toMillis(maxInactiveIntervalInSeconds) >= lastAccessedTime));
+		}
+
+		public  void setLastAccessedTime(long lastAccessedTime) {
+			this.lastAccessedTime = lastAccessedTime;
+		}
+
+		@Override
+		public long getLastAccessedTime() {
+			return lastAccessedTime;
+		}
+
+		@Override
+		public void setMaxInactiveIntervalInSeconds(final int interval) {
+			this.maxInactiveIntervalInSeconds = interval;
+		}
+
+		@Override
+		public int getMaxInactiveIntervalInSeconds() {
+			return maxInactiveIntervalInSeconds;
+		}
+
+		public void setPrincipalName(String principalName) {
+			setAttribute(Session.PRINCIPAL_NAME_ATTRIBUTE_NAME, principalName);
+		}
+
+		public String getPrincipalName() {
+			return getAttribute(Session.PRINCIPAL_NAME_ATTRIBUTE_NAME);
+		}
+
+		@Override
+		public void toData(PdxWriter writer) {
+			writer.writeString("id", getId());
+			writer.writeLong("creationTime", getCreationTime());
+			writer.writeLong("lastAccessedTime", getLastAccessedTime());
+			writer.writeInt("maxInactiveIntervalInSeconds", getMaxInactiveIntervalInSeconds());
+			writer.writeString("principalName", getPrincipalName());
+
+			Set<String> attributeNames = new HashSet<String>(nullSafeSet(getAttributeNames()));
+
+			writer.writeObject("attributeNames", attributeNames);
+
+			for (String attributeName : attributeNames) {
+				writer.writeObject(attributeName, getAttribute(attributeName));
+			}
```

Comment:
I'm wondering if this is writing every attribute every time. If so, is there a way to only write the updates? This will increase performance a lot and will handle race conditions better. For example:

* Assume we have a Session with the attribute of username
* Thread1 - starts to process a request to get a protected resource. The username attribute is present in the session.
* Thread2 - processes a request to log the user out (remove the username attribute).
* Thread2 - The username attribute is removed from the session and Redis is updated to reflect the username is removed
* Thread1 - the request finishes and the Session is updated to include all of the original Session attributes. This means the username is now set in the Session which indicates the user is logged in. However, the user should be logged out.

If we were to write only the delta, this would not be an issue. However, writing every attribute every time alleviates problems like #311